base/special/gamma.jl

```diff
@@ -404,13 +404,124 @@ invdigamma(x::Float32) = Float32(invdigamma(Float64(x)))
 invdigamma(x::Real) = invdigamma(Float64(x))
 @vectorize_1arg Real invdigamma
 
-function beta(x::Number, w::Number)
-    yx, sx = lgamma_r(x)
-    yw, sw = lgamma_r(w)
-    yxw, sxw = lgamma_r(x+w)
-    return exp(yx + yw - yxw) * (sx*sw*sxw)
+const MAXGAM = 171.624376956302725
+const ASYMP_FACTOR = 1e6
+
+function beta(a::Number, b::Number)
+    real(a) <= 0.0 && isinteger(a) && return beta_negint(a, b)
+    real(b) <= 0.0 && isinteger(b) && return beta_negint(b, a)
+
+    if abs(a) < abs(b)
```

Comment:
Note that `abs` is unnecessarily slow for complex numbers.  I would tend to define a `fastabs(x)`, where `fastabs(x::Real) = abs(x)` and `fastabs(z::Complex) = abs(real(z)) + abs(imag(z))`.   This is within a factor of `sqrt(2)` of `abs(z)` but is much faster, and the `sqrt(2)` factor doesn't really matter for an application like this if you adjust the thresholds accordingly.